File: src/lib/api/services.ts
import { invoke } from '@/lib/api/invoke'

export interface Postgres {
  /** Cluster directory, binaries and data. */
  dir: string
  /** Loopback port the cluster listens on. */
  port: number
}

/** Start the one Postgres cluster, initializing it on first call. */
export function postgresEnsure(): Promise<Postgres> {
  return invoke<Postgres>('services_postgres_ensure')
}

/** Stop the cluster. A cluster already down resolves without doing anything. */
export function postgresStop(): Promise<Postgres> {
  return invoke<Postgres>('services_postgres_stop')
}

/**
 * Create the database and role for one subscribed app, and resolve with its
 * DATABASE_URL. The cluster is derived in Rust; the frontend names the app.
 */
export function provisionAppDb(appName: string): Promise<string> {
  return invoke<string>('services_provision_app_db', { appName })
}

/**
 * Apply one subscribed app's declared migrations to its own database.
 *
 * The folder comes from the manifest of the clone on disk, not from here, so
 * there is no path to pass and none to get wrong.
 */
export function applyMigrations(appName: string): Promise<number> {
  return invoke<number>('services_apply_migrations', { appName })
}

File: src-tauri/src/commands/services.rs
//! Commands over `aias_core::services`.
//!
//! The frontend names an app and nothing else. It cannot hand in a DSN, a
//! cluster or a directory: those are derived here from `paths::data_dir` and
//! from the app's own manifest, so a compromised web context cannot point the
//! platform at another database or another folder of SQL.

use aias_core::services::{self, Postgres};
use aias_core::{apps, paths};

use crate::error::{CmdError, CmdResult};

#[tauri::command]
pub async fn services_postgres_ensure() -> CmdResult<Postgres> {
    Ok(services::ensure_postgres(&paths::data_dir()).await?)
}

/// Create the database and role for one subscribed app, and give back its DSN.
#[tauri::command]
pub async fn services_provision_app_db(app_name: String) -> CmdResult<String> {
    let data_dir = paths::data_dir();
    let pg = services::ensure_postgres(&data_dir).await?;
    Ok(services::provision_app_db(&pg, &app_name).await?)
}

/// Apply one subscribed app's declared migrations to its own database.
///
/// The folder comes from the manifest of the clone on disk and is resolved
/// against the app directory, so it cannot be pointed anywhere else.
#[tauri::command]
pub async fn services_apply_migrations(app_name: String) -> CmdResult<usize> {
    let data_dir = paths::data_dir();
    let app = apps::installed(&data_dir)
        .into_iter()
        .find(|app| app.name == app_name)
        .ok_or_else(|| CmdError {
            code: "not_found".into(),
            message: format!("app `{app_name}` is not subscribed"),
        })?;
    let Some(postgres) = app.manifest.services.postgres else {
        return Ok(0);
    };

    let dir = apps::resolve_migrations(&app.dir, &postgres.migrations)?;
    let pg = services::ensure_postgres(&data_dir).await?;
    let database_url = services::provision_app_db(&pg, &app_name).await?;
    Ok(services::apply_migrations(&database_url, &dir).await?)
}

#[tauri::command]
pub async fn services_postgres_stop() -> CmdResult<Postgres> {
    let pg = services::postgres_at(&paths::data_dir());
    services::stop_postgres(&pg).await?;
    Ok(pg)
}
